### PR TITLE
Add experimental sparse CUDA SpikeLinear kernels

### PR DESCRIPTION
**Pull request by AllenYolk**
## Summary

- add experimental hand-written CUDA SpikeLinear kernels compiled through CuPy/NVRTC
- register packed-dense and row-sparse custom ops with fake implementations and autograd formulas
- expose `sparse_linear` with explicit `torch` and `sparse` strategies; default to the safe PyTorch implementation and perform no density synchronization
- keep v3 as a low-level custom op for caller-owned pre-packed `uint8` spikes
- support FP32, FP16, and BF16 inputs in the bit-pack utility
- add CUDA safety validation, device/stream handling, per-device kernel caches, API docs, and focused tests

## Scope

This PR proposes experimental kernels and kernel-level evidence only. It does not claim end-to-end model speedup or automatic production dispatch. `sparse` is opt-in and intended for workloads profiled by the caller; dense and higher-density cases can be slower than cuBLAS.

## Validation

- A100: `73 passed, 1 warning`
- RTX 4090: `73 passed, 1 warning`
- formatter, Ruff, compileall, `git diff --check`, and the new Sphinx API page pass
- first backward may emit PyTorch's one-time cuBLAS primary-context warning after CuPy initializes CUDA; correctness is unaffected

The benchmark results below were generated from implementation commit `e78a0f99`. The review follow-up renamed public symbols and extended bit packing; the timed row-sparse CUDA implementation is unchanged. Raw benchmark JSON is intentionally not checked in.

Representative sparse-kernel speedups versus dense `F.linear`:

| GPU | Density | 1024^3 | 2048^3 | 4096^3 |
| --- | ---: | ---: | ---: | ---: |
| A100 | 0.02 | 1.62x | 2.84x | 1.82x |
| A100 | 0.05 | 1.39x | 1.85x | 1.22x |
| RTX 4090 | 0.02 | 1.10x | 1.80x | 2.09x |

End-to-end integration and model-level performance evaluation are intentionally deferred to follow-up work.

## Summary by CodeRabbit

* **New Features**
  * Added experimental CUDA-accelerated binary spike linear operations with dense and sparse execution strategies.
  * Added bit-packing support for spike tensors and exposed the new APIs publicly.
  * Added API documentation for the spike linear module.

* **Performance**
  * Added benchmarks comparing dense and sparse execution across matrix sizes and densities, with timing, speedup, and correctness results.

* **Tests**
  * Added comprehensive coverage for numerical accuracy, gradients, validation, devices, streams, edge cases, and compilation compatibility.

### CODE REVIEW

File: spikingjelly/activation_based/cuda_kernel/spike_linear.py [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+# ----------------------------------------------------------------------
+
+
+def bit_pack_spike_dense(spike: torch.Tensor) -> torch.Tensor:
+    """Pack a contiguous binary FP32 CUDA matrix into row-major uint8 bytes.
+
+    Bit ``i`` of byte ``b`` represents ``spike[:, b * 8 + i]``. A trailing
+    partial byte is zero-padded. Values greater than ``0.5`` are treated as 1.
+    """
+    _check_cuda_tensor(spike, "spike", torch.float32, 2)
```
Comment (AllenYolk) [inline spikingjelly/activation_based/cuda_kernel/spike_linear.py]: Is it possible to make bit-pack agnostic to input dtype? Can we make fp16, bf16 compatible in this case?

File: spikingjelly/activation_based/cuda_kernel/spike_linear.py [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+    _check_cuda_tensor(spike_packed, "spike_packed", torch.uint8, 2)
+    if weight.dim() != 2:
+        raise ValueError("weight must be 2D")
+    M, K_PACKED = spike_packed.shape
+    N, K = weight.shape
+    _check_weight_bias(weight, bias, K, spike_packed.device)
+    if K_PACKED != (K + 7) // 8:
+        raise ValueError("spike_packed.shape[1] must equal ceil(weight.shape[1] / 8)")
+    if M * N > _MAX_CUDA_ELEMENTS:
+        raise ValueError("output exceeds the CUDA kernel element limit")
```
Comment (AllenYolk) [inline spikingjelly/activation_based/cuda_kernel/spike_linear.py:388]: Defensive coding is not preferred. Remove unnecessary runtime checks.

File: spikingjelly/activation_based/cuda_kernel/spike_linear.py [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+# ----------------------------------------------------------------------
+# Public API: sparse_linear
+# ----------------------------------------------------------------------
+
+
+def sparse_linear(
+    spike: torch.Tensor,
+    weight: torch.Tensor,
+    bias: Optional[torch.Tensor] = None,
+    strategy: Literal["torch", "v15_sparse"] = "torch",
```
Comment (AllenYolk) [inline spikingjelly/activation_based/cuda_kernel/spike_linear.py]: "v15" should not be exposed to users. It's just a version number for internal use!
